src/main/java/io/supertokens/multitenancy/MultitenancyHelper.java

```diff
@@ -84,80 +85,85 @@ private TenantConfig[] getAllTenantsFromDb() throws StorageQueryException {
         return StorageLayer.getMultitenancyStorage(main).getAllTenants();
     }
 
-    public void refreshTenantsInCoreIfRequired(boolean reloadAllResources) {
+    public List<TenantIdentifier> refreshTenantsInCoreIfRequired(boolean reloadAllResources) {
         try {
-            main.getResourceDistributor().withResourceDistributorLock(() -> {
+            return main.getResourceDistributor().withResourceDistributorLock(() -> {
                 try {
                     TenantConfig[] tenantsFromDb = getAllTenantsFromDb();
 
-                    boolean hasChanged = false;
-                    if (tenantsFromDb.length != tenantConfigs.length) {
-                        hasChanged = true;
-                    } else {
-                        Map<TenantIdentifier, TenantConfig> fromDb = new HashMap<>();
-                        for (TenantConfig t : tenantsFromDb) {
-                            fromDb.put(t.tenantIdentifier, t);
-                        }
-                        for (TenantConfig t : this.tenantConfigs) {
-                            TenantConfig fromDbConfig = fromDb.get(t.tenantIdentifier);
-                            if (!t.deepEquals(fromDbConfig)) {
-                                hasChanged = true;
-                                break;
-                            }
+                    Map<ResourceDistributor.KeyClass, JsonObject> normalizedTenantsFromDb = Config.getNormalisedConfigsForAllTenants(
+                            tenantsFromDb, Config.getBaseConfigAsJsonObject(main));
+
+                    Map<ResourceDistributor.KeyClass, JsonObject> normalizedTenantsFromMemory = Config.getNormalisedConfigsForAllTenants(
+                            this.tenantConfigs, Config.getBaseConfigAsJsonObject(main));
+
+                    List<TenantIdentifier> tenantsThatChanged = new ArrayList<>();
+
+                    for (Map.Entry<ResourceDistributor.KeyClass, JsonObject> entry : normalizedTenantsFromMemory.entrySet()) {
+                        JsonObject tenantConfigFromMemory = entry.getValue();
+                        JsonObject tenantConfigFromDb = normalizedTenantsFromDb.get(entry.getKey());
+
+                        if (!tenantConfigFromMemory.equals(tenantConfigFromDb)) {
+                            tenantsThatChanged.add(entry.getKey().getTenantIdentifier());
                         }
                     }
 
+                    boolean sameNumberOfTenants = tenantsFromDb.length == this.tenantConfigs.length;
+
                     this.tenantConfigs = tenantsFromDb;
-                    if (!hasChanged) {
-                        return;
+                    if (tenantsThatChanged.size() == 0 && sameNumberOfTenants) {
+                        return tenantsThatChanged;
                     }
 
```

Comment:
add a process state here and in the test for when we check that there are not unnecessary reloads, check that process state is null

Reply:
Done